### DISCUSSION TITLE
How to share an instance between the Java Agent and the target Application

### DISCUSSION THREAD

**Discussion by adinauer** [open-telemetry/opentelemetry-java-instrumentation#7042] [Q&A] [answered]
Hello 👋 

I'm trying to access a static field of the target application from a `SpanProcessor` which I added to an extension and then exported as an agent JAR as described [in your README](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/README.md).

It seems it's not the same class when I access it from within the target application and when accessing it from the `SpanProcessor` that I supplied as an agent extension. In the `SpanProcessor` it gives me `ExtensionClassLoader` while in the target application it's `ClassLoaders$AppClassLoader` when checking the `classLoader` of the class.

I've tried adding a `TypeInstrumentation` with an `@Advice.OnMethodExit()`. In that advice I can access the static field and also print it using `System.out.println`. But when I try to assign that to a field on e.g. the `TypeInstrumentation` it seems to no longer be running the advice as I can no longer see any of the `println`s. (Side question: can I somehow turn on more logging for this problem?)

What's the intended way of accessing a static field or sharing an instance of an object between the agent extension and the target application?

**Comment by laurit** [comment, marked as answer]
Application and extension are loaded in different class loaders, you can not directly read a field in application code from an extension. To share an object you'll have to place it in a class loader that is visible to both the application and the extension, that would be the boot loader. Unfortunately this currently can't be easily done with an extension. There is a sample for it in the custom distro, see https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/distro/bootstrap/src/main/java/com/example/javaagent/bootstrap. Alternatively if the application class that you wish to access is accessible though system class loader you could use `Class.forName` with system class loader and reflection to get access to it. If the value that you want to get access to is something simple e.g. String you could use `System.setProperty` in application code (this call could be added with instrumentation) to expose it to the extension.
If you don't see your debug printlns after modifying instrumentation it most likely means that there is an error in your instrumentation that prevents applying it. You can enable debug logging with `-Dotel.javaagent.debug=true`

**Comment by adinauer** [comment]
Thanks @laurit!

I've tried using reflection but that would mean performing every interaction with the shared instance (and any nested objects) through reflection so I will now switch to creating a custom distro as shown in your example.

**Reply by jianwu** [reply]
Reflection works for me. 

`ClassLoader.getSystemClassLoader().loadClass("..."). `
